File: src/test/java/com/example/librarymanagement/service/impl/LibraryServiceImplTest.java
package com.example.librarymanagement.service.impl;

import com.example.librarymanagement.data.TestDataFactory;
import com.example.librarymanagement.model.entity.Library;
import com.example.librarymanagement.repository.LibraryRepository;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class LibraryServiceImplTest {

    @InjectMocks
    private LibraryServiceImpl libraryServiceImpl;

    @Mock
    private LibraryRepository libraryRepository;
    @Mock
    private BookPresenceServiceImpl bookPresenceServiceImpl;
    @Mock
    private ReservationServiceImpl reservationService;

    private static Library library1;

    @BeforeAll
    public static void init() {
        TestDataFactory.TestDataRel testDataRel = TestDataFactory.createTestDataRel();

        library1 = testDataRel.library;
    }

    @Test
    public void testFindAll() {
        when(libraryRepository.findAll()).thenReturn(List.of(library1));

        assertEquals(List.of(library1), libraryServiceImpl.findAll());
        verify(libraryRepository, times(1)).findAll();
    }

    @Test
    public void testGetLibraryById() {
        when(libraryRepository.findById(anyLong())).thenReturn(Optional.of(library1));

        assertEquals(library1, libraryServiceImpl.getLibraryById(1L));
        verify(libraryRepository, times(1)).findById(1L);
    }

    @Test
    public void testCreateLibrary() {
        when(libraryRepository.save(library1)).thenReturn(library1);

        assertEquals(library1, libraryServiceImpl.createLibrary(library1));
        verify(libraryRepository, times(1)).save(library1);
    }

    @Test
    public void testUpdateLibrary() {
        when(libraryRepository.findById(anyLong())).thenReturn(Optional.of(library1));
        when(libraryRepository.save(library1)).thenReturn(library1);

        assertEquals(library1, libraryServiceImpl.updateLibrary(1L, library1));
        verify(libraryRepository, times(1)).findById(1L);
        verify(libraryRepository, times(1)).save(library1);
    }

    @Test
    public void deleteLibrary() {
        when(libraryRepository.findById(1L)).thenReturn(Optional.of(library1));
        doNothing().when(libraryRepository).deleteById(1L);

        libraryServiceImpl.deleteLibrary(1L);
        verify(libraryRepository, times(1)).deleteById(1L);
    }
}

File: src/test/java/com/example/librarymanagement/service/impl/LibraryServiceImplTest.kt
package com.example.librarymanagement.service.impl

import com.example.librarymanagement.data.LibraryDataFactory
import com.example.librarymanagement.data.ReservationDataFactory
import com.example.librarymanagement.repository.LibraryRepository
import com.example.librarymanagement.service.BookPresenceService
import com.example.librarymanagement.service.ReservationService
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class LibraryServiceImplTest {
    private val libraryRepository: LibraryRepository = mockk()
    private val bookPresenceService: BookPresenceService = mockk()
    private val reservationService: ReservationService = mockk()
    private val libraryService = LibraryServiceImpl(
        libraryRepository
    )
    private val id = "1"

    private val library = LibraryDataFactory.createLibrary(id)

    @Test
    fun shouldCreateLibrary() {
        //GIVEN
        val expected = library

        every { libraryRepository.save(library) } returns library

        //WHEN
        val actual = libraryService.createLibrary(library)

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { libraryRepository.save(library) }
    }

    @Test
    fun shouldUpdateLibrary() {
        //GIVEN
        val expected = library.copy(name = "Updated")

        every { libraryRepository.findById(id) } returns library
        every { libraryRepository.save(expected) } returns expected

        //WHEN
        val actual = libraryService.updateLibrary(expected)

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { libraryRepository.findById(id) }
        verify(exactly = 1) { libraryRepository.save(expected) }
    }

    @Test
    fun shouldDeleteLibraryById() {
        //GIVEN
        val reservation = ReservationDataFactory.createReservation(id)

        every { libraryRepository.findById(id) } returns library
        every { bookPresenceService.deleteBookPresenceById(id) } returns Unit
        every { reservationService.getReservationsByUserId(id) } returns listOf(reservation)
        every { reservationService.deleteReservationById(id) } returns Unit
        every { libraryRepository.deleteById(id) } returns Unit

        //WHEN
        libraryService.deleteLibraryById(id)

        //THEN
        verify(exactly = 1) { libraryRepository.deleteById(id) }
    }

    @Test
    fun shouldGetLibraryById() {
        //GIVEN
        val expected = library
        every { libraryRepository.findById(id) } returns library

        //WHEN
        val actual = libraryService.getLibraryById(id)

        //THEN
        Assertions.assertEquals(expected, actual)
        verify(exactly = 1) { libraryRepository.findById(id) }
    }

    @Test
    fun shouldFindAll() {
        //GIVEN
        val excepted = listOf(library)
        every { libraryRepository.findAll() } returns listOf(library)

        //WHEN
        val result = libraryService.findAll()

        //THEN
        Assertions.assertEquals(excepted, result)
        verify(exactly = 1) { libraryRepository.findAll() }
    }
}
